pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -492,19 +492,35 @@ def run(
 
                 .. _docs: https://bit.ly/3il9A2u
         """
-        if "max_step" not in options:
-            pulse_durations = [
-                slot.tf - slot.ti
-                for ch_sample in self.samples_obj.samples_list
-                for slot in ch_sample.slots
-                if not (
-                    np.all(np.isclose(ch_sample.amp[slot.ti : slot.tf], 0))
-                    and np.all(np.isclose(ch_sample.det[slot.ti : slot.tf], 0))
+
+        def get_min_variation(ch_sample: ChannelSamples) -> int:
+            end_point = ch_sample.duration - 1
+            min_variations: list[int] = []
+            for sample in (ch_sample.amp, ch_sample.det):
+                # Find the indices at which the sample varies
+                variations_indices: np.ndarray = np.concatenate(
+                    [
+                        np.array([-1]),
+                        np.nonzero(np.diff(sample)),
+                        np.array([end_point]),
+                    ],
+                    axis=None,
                 )
-            ]
-            if pulse_durations:
-                options["max_step"] = 0.5 * min(pulse_durations) / 1000
+                # Find the differences between two indices at which sample varies
+                min_variations.append(int(np.min(np.diff(variations_indices))))
+
+            return min(min_variations)
```

Comment:
I have a feeling we are going to get `max_step=0.001` whenever we have something other than constant waveforms in the sequence. Hopefully that doesn't slow us down too much

Reply:
I agree that robustness is preferable and speed can still be recovered by manually setting `max_step`. 
But I also think it's a good idea to have a notion of how much this impacts speed at the moment, if things become significantly slower the users won't like it.

Reply:
I was considering looking at the slots and checking what is the type of the waveform used in each Pulse slot. If the slot is a CompositeWaveform/CustomWaveform/InterpolatedWaveform then it's best to define it using diff, otherwise we can say that max_step should be half this slot as it was before.
However, this loop might take more time than the simulation in itself...

Reply:
This sounds like overkill to me

Reply:
To me as well haha